In [26]:
import pandas as pd 
import matplotlib.pyplot as plt

In [29]:
df_ALLCountries_Data = pd.read_csv('./IDS_CSV/IDS_ALLCountries_Data.csv')
df_ALLCountries_Data

Country Name Country Code  \
0                                            Afghanistan          AFG   
1                                            Afghanistan          AFG   
2                                            Afghanistan          AFG   
3                                            Afghanistan          AFG   
4                                            Afghanistan          AFG   
...                                                  ...          ...   
77384                                                NaN          NaN   
77385                                                NaN          NaN   
77386                                                NaN          NaN   
77387  Data from database: International Debt Statistics          NaN   
77388                           Last Updated: 10/10/2021          NaN   

      Counterpart-Area Name Counterpart-Area Code  \
0                     World                   WLD   
1                     World                   WLD   
2                     World                   WLD   
3                     World                   WLD   
4                     World                   WLD   
...                     ...                   ...   
77384                   NaN                   NaN   
77385                   NaN                   NaN   
77386                   NaN                   NaN   
77387                   NaN                   NaN   
77388                   NaN                   NaN   

                                             Series Name  Series Code  1970  \
0      Average grace period on new external debt comm...  DT.GPA.DPPG   NaN   
1      Average grace period on new external debt comm...  DT.GPA.OFFT   NaN   
2      Average grace period on new external debt comm...  DT.GPA.PRVT   NaN   
3      Average grant element on new external debt com...  DT.GRE.DPPG   NaN   
4      Average grant element on new external debt com...  DT.GRE.OFFT   NaN   
...                                                  ...          ...   ...   
77384                                                NaN          NaN   NaN   
77385                                                NaN          NaN   NaN   
77386                                                NaN          NaN   NaN   
77387                                                NaN          NaN   NaN   
77388                                                NaN          NaN   NaN   

       1971  1972  1973  ...     2019  2020  2021  2022  2023  2024  2025  \
0       NaN   NaN   NaN  ...  17.9041   0.0   NaN   NaN   NaN   NaN   NaN   
1       NaN   NaN   NaN  ...  17.9041   0.0   NaN   NaN   NaN   NaN   NaN   
2       NaN   NaN   NaN  ...   0.0000   0.0   NaN   NaN   NaN   NaN   NaN   
3       NaN   NaN   NaN  ...  73.0790   0.0   NaN   NaN   NaN   NaN   NaN   
4       NaN   NaN   NaN  ...  73.0790   0.0   NaN   NaN   NaN   NaN   NaN   
...     ...   ...   ...  ...      ...   ...   ...   ...   ...   ...   ...   
77384   NaN   NaN   NaN  ...      NaN   NaN   NaN   NaN   NaN   NaN   NaN   
77385   NaN   NaN   NaN  ...      NaN   NaN   NaN   NaN   NaN   NaN   NaN   
77386   NaN   NaN   NaN  ...      NaN   NaN   NaN   NaN   NaN   NaN   NaN   
77387   NaN   NaN   NaN  ...      NaN   NaN   NaN   NaN   NaN   NaN   NaN   
77388   NaN   NaN   NaN  ...      NaN   NaN   NaN   NaN   NaN   NaN   NaN   

       2026  2027  2028  
0       NaN   NaN   NaN  
1       NaN   NaN   NaN  
2       NaN   NaN   NaN  
3       NaN   NaN   NaN  
4       NaN   NaN   NaN  
...     ...   ...   ...  
77384   NaN   NaN   NaN  
77385   NaN   NaN   NaN  
77386   NaN   NaN   NaN  
77387   NaN   NaN   NaN  
77388   NaN   NaN   NaN  

[77389 rows x 65 columns]

In [33]:
df_ALLCountries_Data.columns

Index(['Country Name', 'Country Code', 'Counterpart-Area Name',
       'Counterpart-Area Code', 'Series Name', 'Series Code', '1970', '1971',
       '1972', '1973', '1974', '1975', '1976', '1977', '1978', '1979', '1980',
       '1981', '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989',
       '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998',
       '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
       '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
       '2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024', '2025',
       '2026', '2027', '2028'],
      dtype='object')

In [28]:
df_ALLCountries_Data.shape

(77389, 65)

### Finding the number of distinct countries

In [49]:
# item_counts = df["Your_Column"].value_counts()
number = df_ALLCountries_Data['Country Name'].value_counts()
number

Afghanistan                                          569
Panama                                               569
Nepal                                                569
Nicaragua                                            569
Niger                                                569
                                                    ... 
Grenada                                              569
Guatemala                                            569
Guinea                                               569
Data from database: International Debt Statistics      1
Last Updated: 10/10/2021                               1
Name: Country Name, Length: 138, dtype: int64

### The number of distinct countries are 138 as given by the length 

### Below we will find the names of the countries 

In [48]:
df_ALLCountries_Data['Country Name'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Armenia', 'Azerbaijan', 'Bangladesh', 'Belarus', 'Belize',
       'Benin', 'Bhutan', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana',
       'Brazil', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cabo Verde',
       'Cambodia', 'Cameroon', 'Central African Republic', 'Chad',
       'China', 'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Costa Rica', "Cote d'Ivoire", 'Djibouti', 'Dominica',
       'Dominican Republic',
       'East Asia & Pacific (excluding high income)', 'Ecuador',
       'Egypt, Arab Rep.', 'El Salvador', 'Eritrea', 'Eswatini',
       'Ethiopia', 'Europe & Central Asia (excluding high income)',
       'Fiji', 'Gabon', 'Gambia, The', 'Georgia', 'Ghana', 'Grenada',
       'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti',
       'Honduras', 'IDA only', 'India', 'Indonesia', 'Iran, Islamic Rep.',
       'Jamaica', 'Jordan', 'Kazakhstan', 'Kenya', 'Kosovo',
       'Kyrgyz Repu

In [16]:
df_CountryMetaData = pd.read_csv('./IDS_CSV/IDS_CountryMetaData.csv', encoding='unicode_escape')
df_CountryMetaData.head()

Code                                Long Name         Income Group  \
0  AFG             Islamic State of Afghanistan           Low income   
1  ALB                      Republic of Albania  Upper middle income   
2  DZA  People's Democratic Republic of Algeria  Lower middle income   
3  AGO              People's Republic of Angola  Lower middle income   
4  ARG                       Argentine Republic  Upper middle income   

                       Region Lending category Other groups   Currency Unit  \
0                  South Asia              IDA         HIPC  Afghan afghani   
1       Europe & Central Asia             IBRD          NaN    Albanian lek   
2  Middle East & North Africa             IBRD          NaN  Algerian dinar   
3          Sub-Saharan Africa             IBRD          NaN  Angolan kwanza   
4   Latin America & Caribbean             IBRD          NaN  Argentine peso   

  Latest population census                     Latest household survey  \
0                     1979         Demographic and Health Survey, 2015   
1          2020 (expected)      Demographic and Health Survey, 2017/18   
2          2020 (expected)  Multiple Indicator Cluster Survey, 2018/19   
3                     2014      Demographic and Health Survey, 2015/16   
4          2020 (expected)  Multiple Indicator Cluster Survey, 2019/20   

                                       Special Notes  ...  \
0  Fiscal year end: March 20; reporting period fo...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

   Source of most recent Income and expenditure data  \
0         Integrated household survey (IHS), 2016/17   
1  Living Standards Measurement Study Survey (LSM...   
2            Integrated household survey (IHS), 2011   
3         Integrated household survey (IHS), 2008/09   
4            Integrated household survey (IHS), 2016   

   Vital registration complete Latest agricultural census  \
0                          NaN                        NaN   
1                          Yes                       2012   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          Yes                       2008   

  Latest industrial data Latest trade data Latest water withdrawal data  \
0                    NaN            2018.0                       2000.0   
1                 2013.0            2018.0                       2006.0   
2                 2010.0            2017.0                       2012.0   
3                    NaN            2018.0                       2005.0   
4                 2002.0            2018.0                       2011.0   

  2-alpha code WB-2 code   Table Name   Short Name  
0           AF        AF  Afghanistan  Afghanistan  
1           AL        AL      Albania      Albania  
2           DZ        DZ      Algeria      Algeria  
3           AO        AO       Angola       Angola  
4           AR        AR    Argentina    Argentina  

[5 rows x 31 columns]

In [54]:
df_CountryMetaData.columns

Index(['Code', 'Long Name', 'Income Group', 'Region', 'Lending category',
       'Other groups', 'Currency Unit', 'Latest population census',
       'Latest household survey', 'Special Notes',
       'National accounts base year', 'National accounts reference year',
       'System of National Accounts', 'SNA price valuation',
       'Alternative conversion factor', 'PPP survey years',
       'Balance of Payments Manual in use', 'External debt Reporting status',
       'System of trade', 'Government Accounting concept',
       'IMF data dissemination standard',
       'Source of most recent Income and expenditure data',
       'Vital registration complete', 'Latest agricultural census',
       'Latest industrial data', 'Latest trade data',
       'Latest water withdrawal data', '2-alpha code', 'WB-2 code',
       'Table Name', 'Short Name'],
      dtype='object')

In [18]:
df_Country_SeriesMetaData = pd.read_csv('./IDS_CSV/IDS_Country-SeriesMetaData.csv', encoding='unicode_escape')
df_Country_SeriesMetaData.head()

Type       Country Code  \
0  Country-Series  Afghanistan (AFG)   
1  Country-Series  Afghanistan (AFG)   
2  Country-Series       Angola (AGO)   
3  Country-Series       Angola (AGO)   
4  Country-Series      Albania (ALB)   

                                         Series Code  \
0  Foreign direct investment, net inflows in repo...   
1  External debt stocks, total (DOD, current US$)...   
2  Foreign direct investment, net inflows in repo...   
3  External debt stocks, total (DOD, current US$)...   
4  Foreign direct investment, net inflows in repo...   

                                         Description  
0  Data on Foreign Direct Investment are based on...  
1  Long-term public and publicly guaranteed and l...  
2  Data on Foreign Direct Investment are based on...  
3  Long-term public and publicly guaranteed debt ...  
4  Data on Foreign Direct Investment are based on...

In [55]:
df_Country_SeriesMetaData.columns

Index(['Type', 'Country Code', 'Series Code', 'Description'], dtype='object')

In [20]:
df_Footnotes = pd.read_csv('./IDS_CSV/IDS_Footnotes.csv', encoding='unicode_escape')
df_Footnotes.head()

Type       Country Code  \
0  FootNote  Afghanistan (AFG)   
1  FootNote  Afghanistan (AFG)   
2  FootNote  Afghanistan (AFG)   
3  FootNote  Afghanistan (AFG)   
4  FootNote  Afghanistan (AFG)   

                                         Series Code      Time Code  \
0  Foreign direct investment, net inflows in repo...  1970 (YR1970)   
1  Foreign direct investment, net inflows in repo...  1971 (YR1971)   
2  Foreign direct investment, net inflows in repo...  1972 (YR1972)   
3  Foreign direct investment, net inflows in repo...  1973 (YR1973)   
4  Foreign direct investment, net inflows in repo...  1976 (YR1976)   

                                         Description  
0  Source: United Nations Conference on Trade and...  
1  Source: United Nations Conference on Trade and...  
2  Source: United Nations Conference on Trade and...  
3  Source: United Nations Conference on Trade and...  
4  Source: United Nations Conference on Trade and...

In [56]:
df_Footnotes.columns

Index(['Type', 'Country Code', 'Series Code', 'Time Code', 'Description'], dtype='object')

In [51]:
df_SeriesMetaData = pd.read_csv('./IDS_CSV/IDS_SeriesMetaData.csv', encoding='unicode_escape')
df_SeriesMetaData.head()

Code License Type  \
0  DT.GPA.DPPG          NaN   
1  DT.GPA.OFFT          NaN   
2  DT.GPA.PRVT          NaN   
3  DT.GRE.DPPG          NaN   
4  DT.GRE.OFFT          NaN   

                                      Indicator Name  \
0  Average grace period on new external debt comm...   
1  Average grace period on new external debt comm...   
2  Average grace period on new external debt comm...   
3  Average grant element on new external debt com...   
4  Average grant element on new external debt com...   

                                    Short definition  \
0  Grace period is the period from the date of si...   
1  Grace period is the period from the date of si...   
2  Grace period is the period from the date of si...   
3  The grant element of a loan is the grant equiv...   
4  The grant element of a loan is the grant equiv...   

                                     Long definition  \
0  Grace period is the period from the date of si...   
1  Grace period is the period from the date of si...   
2  Grace period is the period from the date of si...   
3  The grant element of a loan is the grant equiv...   
4  The grant element of a loan is the grant equiv...   

                                       Source  \
0  World Bank, International Debt Statistics.   
1  World Bank, International Debt Statistics.   
2  World Bank, International Debt Statistics.   
3  World Bank, International Debt Statistics.   
4  World Bank, International Debt Statistics.   

                                          Topic  \
0  Economic Policy & Debt: External debt: Terms   
1  Economic Policy & Debt: External debt: Terms   
2  Economic Policy & Debt: External debt: Terms   
3  Economic Policy & Debt: External debt: Terms   
4  Economic Policy & Debt: External debt: Terms   

                         Dataset Periodicity Aggregation method  \
0  International Debt Statistics      Annual   Weighted average   
1  International Debt Statistics      Annual   Weighted average   
2  International Debt Statistics      Annual   Weighted average   
3  International Debt Statistics      Annual   Weighted average   
4  International Debt Statistics      Annual   Weighted average   

  General comments  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN

In [57]:
df_SeriesMetaData.columns

Index(['Code', 'License Type', 'Indicator Name', 'Short definition',
       'Long definition', 'Source', 'Topic', 'Dataset', 'Periodicity',
       'Aggregation method', 'General comments'],
      dtype='object')

### Below is finding out the distinct Indicator Names

In [52]:
df_SeriesMetaData['Indicator Name'].unique()

array(['Average grace period on new external debt commitments (years)',
       'Average grace period on new external debt commitments, official (years)',
       'Average grace period on new external debt commitments, private (years)',
       'Average grant element on new external debt commitments (%)',
       'Average grant element on new external debt commitments, official (%)',
       'Average grant element on new external debt commitments, private (%)',
       'Average interest on new external debt commitments (%)',
       'Average interest on new external debt commitments, official (%)',
       'Average interest on new external debt commitments, private (%)',
       'Average maturity on new external debt commitments (years)',
       'Average maturity on new external debt commitments, official (years)',
       'Average maturity on new external debt commitments, private (years)',
       'CB, bilateral (AMT, current US$)',
       'CB, bilateral (DIS, current US$)',
       'CB, bilater

### Finding out number of distinct Indicators

In [53]:
NumberOfIndicators = df_SeriesMetaData['Indicator Name'].value_counts()
NumberOfIndicators

Average grace period on new external debt commitments (years)    1
PPG, multilateral concessional (DOD, current US$)                1
PPG, multilateral (DIS, current US$)                             1
PPG, multilateral (DOD, current US$)                             1
PPG, multilateral (INT, current US$)                             1
                                                                ..
GG, official creditors (TDS, current US$)                        1
GG, official creditors (NTR, current US$)                        1
GG, official creditors (NFL, current US$)                        1
GG, official creditors (INT, current US$)                        1
Use of IMF credit, SDR allocations (DOD, current US$)            1
Name: Indicator Name, Length: 562, dtype: int64

### We see the number of distinct indicators are 562

### Finding out the most common Debt Indicator

In [59]:
df_SeriesMetaData['Indicator Name'].value_counts().idxmax()

'Average grace period on new external debt commitments (years)'

### We see the most common Debt Indicator is 'Average grace period on new external debt commitments (years)'

In [21]:
df_Series_TimeMetaData = pd.read_csv('./IDS_CSV/IDS_Series-TimeMetaData.csv', encoding='unicode_escape')
df_Series_TimeMetaData.head()

Type                                        Series Code  Time Code  \
0  Series-Time  Present value of external debt (current US$) (...   (YR2020)   
1  Series-Time  Present value of external debt (% of exports o...   (YR2020)   
2  Series-Time  Present value of external debt (% of GNI) (DT....   (YR2020)   

                                         Description  
0  Present value calculations for these countries...  
1  Present value calculations for these countries...  
2  Present value calculations for these countries...

In [58]:
df_Series_TimeMetaData.columns

Index(['Type', 'Series Code', 'Time Code', 'Description'], dtype='object')